consensus/src/dag/tests/reliable_broadcast_tests.rs

```diff
@@ -121,33 +107,71 @@ async fn test_reliable_broadcast() {
     let (_, validator_verifier) = random_validator_verifier(5, None, false);
     let validators = validator_verifier.get_ordered_account_addresses();
     let failures = HashMap::from([(validators[0], 1), (validators[2], 3)]);
     let sender = Arc::new(TestDAGSender::new(failures));
     let rb = ReliableBroadcast::new(validators.clone(), sender);
-    let message = TestMessage(vec![1, 2, 3]);
-
-    // explicit send cancel
-    let (tx, rx) = oneshot::channel();
-    let (cancel_tx, cancel_rx) = oneshot::channel();
-    cancel_tx.send(()).unwrap();
-    tokio::spawn(rb.broadcast::<TestBroadcastStatus>(message.clone(), tx, cancel_rx));
-    assert!(rx.await.is_err());
+    let message = TestMessage(vec![42; validators.len()]);
+    let expected = validators.iter().cloned().collect();
+    let aggregating = TestBroadcastStatus {
+        threshold: validators.len(),
+        received: HashSet::new(),
+    };
+    let fut = rb
+        .broadcast::<TestBroadcastStatus>(message.clone(), aggregating)
+        .then(|aggregated| async move {
+            assert_eq!(aggregated, expected);
+            let aggregating = TestBroadcastStatus {
+                threshold: validator_verifier.len(),
+                received: HashSet::new(),
+            };
+            rb.broadcast::<TestBroadcastStatus>(message, aggregating)
+                .await
+        });
+    assert_eq!(fut.await, validators.into_iter().collect());
+}
 
-    // implicit drop cancel
+#[tokio::test]
+async fn test_abort_reliable_broadcast() {
+    let (_, validator_verifier) = random_validator_verifier(5, None, false);
+    let validators = validator_verifier.get_ordered_account_addresses();
+    let failures = HashMap::from([(validators[0], 1), (validators[2], 3)]);
+    let sender = Arc::new(TestDAGSender::new(failures));
+    let rb = ReliableBroadcast::new(validators.clone(), sender);
+    let message = TestMessage(vec![42; validators.len()]);
     let (tx, rx) = oneshot::channel();
-    let (cancel_tx, cancel_rx) = oneshot::channel();
-    drop(cancel_tx);
-    tokio::spawn(rb.broadcast::<TestBroadcastStatus>(message, tx, cancel_rx));
+    let (abort_handle, abort_registration) = AbortHandle::new_pair();
+    let aggregating = TestBroadcastStatus {
+        threshold: validators.len(),
+        received: HashSet::new(),
+    };
+    let fut = Abortable::new(
+        rb.broadcast::<TestBroadcastStatus>(message.clone(), aggregating)
+            .then(|_| async move {
+                let aggregating = TestBroadcastStatus {
+                    threshold: validators.len(),
+                    received: HashSet::new(),
+                };
+                let ret = rb
+                    .broadcast::<TestBroadcastStatus>(message, aggregating)
+                    .await;
+                tx.send(ret)
+            }),
+        abort_registration,
+    );
+    tokio::spawn(fut);
+    abort_handle.abort();
     assert!(rx.await.is_err());
```

Comment:
this might get flaky no? depending on how fast fut can complete, tx.send can happen before abort?

Reply:
I can reorder the abort to before the spawn